plugins/searchonline.koplugin/main.lua

```diff
@@ -0,0 +1,313 @@
+
+    local sanitised_query = url.escape(query)
+    local build_url = chosen_engine:gsub("{q}", sanitised_query)
+
+    if HandleUrlString:isValidUrl(build_url) then
+        Device:openLink(build_url)
+    end
+end
+
+
+function SearchOnline:chooseSearch(query)
+    local radio_buttons = {}
+    for index, engine_url in ipairs(self.search_engines) do
+        table.insert(radio_buttons, { -- build a button for each
+            {                         -- saved search engine
+                text = HandleUrlString:getDomainName(engine_url),
+                provider = engine_url,
+                checked = index == 1
+            }
+        })
+    end
+
+    self.choose_search = RadioButtonWidget:new{
+        title_text = _("Choose search engine"),
+        width_factor = 0.8,
+        radio_buttons = radio_buttons,
+        cancel_text = _("Cancel"),
+        extra_text = _("Remove"),
+        ok_text = _("Search"),
+        callback = function(radio)
+            if radio.provider then
+                self:searchOnline(query, radio.provider)
+            end
+        end,
+        extra_callback = function(radio)
+            if radio.provider then
+                self:removeUserSearchEngine(radio.provider)
+                UIManager:close(self.choose_search)
+                self:chooseSearch(query)
+            end
+        end,
+        cancel_callback = function()
+            UIManager:close(self.choose_search)
+        end
+    }
+    UIManager:show(self.choose_search)
+end
+
+function SearchOnline:addUserSearchEngine()
+    self.user_input = InputDialog:new{
+        title = _("Add a search engine"),
+        input = "",
+        input_hint = _"https://google.com/search?q={q}",
+        description = _("Please add the URL for a search engine in the following format,\n"
+        .. "{q} represents the word you want to search for.\n"
+        .. "https://google.com/search?q={q}\n"
+        .. "https://en.m.wiktionary.org/wiki/{q}"
+        ),
+        buttons = {
+            {
+                {
+                    text = _("Cancel"),
+                    id = "close",
+                    callback = function()
+                        self.user_input:onCloseKeyboard()
+                        UIManager:close(self.user_input)
+                    end,
+                },
+                {
+                    text = _("Save"),
+                    is_enter_default = true,
+                    callback = function()
+                        local input_text = self.user_input:getInputText()
+
+                        -- Check if input is valid
+                        if not input_text or input_text == "" then
+                            UIManager:show(InfoMessage:new{
+                                text = _("Please enter a valid URL"),
+                            })
+                            return
+                        end
+
+                        if true then
+                            -- Check for duplicates

```

Comment:
I guess this `if true` was for testing, should probably be removed and one indentation level spared once tested :)